packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1216 @@
+  /// for the provided vicinity, the method will return null, otherwise, it will
+  /// return the [RenderBox] of the child.
+  ///
+  /// After [layoutChildSequence] is completed, any remaining children that were
+  /// not obtained will be disposed.
+  void layoutChildSequence();
+
+  @override
+  void performLayout() {
+    _childManager.startLayout();
+
+    // Subclass lays out children.
+    layoutChildSequence();
+
+    _validateChildren();
+    _validateContentDimensions();
+    _needsChildRebuild = false;
+    _didResize = false;
+    needsDelegateRebuild = false;
+    assert(_debugOrphans?.isEmpty ?? true);
+    invokeLayoutCallback<BoxConstraints>((BoxConstraints _) {
+      _childManager.endLayout();
+    });
+  }
+
+  // Ensures all children have a layoutOffset, sets paintExtent & paintOffset,
+  // and arranges children in paint order.
+  void _validateChildren() {
+    assert(_leadingXIndex != null);
+    assert(_trailingXIndex != null);
+    assert(_leadingYIndex != null);
+    assert(_trailingYIndex != null);
+    _childrenInPaintOrder.clear();
+    switch (mainAxis) {
+      case Axis.vertical:
+        // Row major traversal.
+        // This seems backwards, but the vertical axis is the typical default
+        // axis for scrolling in Flutter, while Row-major ordering is the
+        // typical default for matrices, which is why the inverse follows
+        // through in the horizontal case below.
+        // Minor
+        for (int minorIndex = _leadingYIndex!; minorIndex <= _trailingYIndex!; minorIndex++) {
+          // Major
+          for (int majorIndex = _leadingXIndex!; majorIndex <= _trailingXIndex!; majorIndex++) {
+            final ChildVicinity vicinity = ChildVicinity(xIndex: majorIndex, yIndex: minorIndex);
+            _checkVicinity(vicinity);
+          }
+        }
+      case Axis.horizontal:
+        // Column major traversal
+        // Minor
+        for (int minorIndex = _leadingXIndex!; minorIndex <= _trailingXIndex!; minorIndex++) {
+          // Major
+          for (int majorIndex = _leadingYIndex!; majorIndex <= _trailingYIndex!; majorIndex++) {
+            final ChildVicinity vicinity = ChildVicinity(xIndex: minorIndex, yIndex: majorIndex);
+            _checkVicinity(vicinity);
+          }
+        }
+    }
+    // Reset for next layout pass.
+    _leadingXIndex = null;
+    _trailingXIndex = null;
+    _leadingYIndex = null;
+    _trailingYIndex = null;
+  }
+
+  void _checkVicinity(ChildVicinity vicinity) {
+    assert(vicinity != ChildVicinity.invalid);
+    // It is possible and valid for an index to be skipped.
+    // For example, a table can have merged cells, spanning multiple
+    // indices, but only represented by one RenderBox.
+    if (_children.containsKey(vicinity)){
+      final RenderBox child = _children[vicinity]!;
+      assert(parentDataOf(child).vicinity == vicinity);
+      updateChildPaintData(child);
+      _childrenInPaintOrder.add(child);
```

Comment:
Establishing the paint order as a second pass after layout seems a little odd. I wonder if we could just store this as a linked list in the parentdata and make layoutChildSequence set this accordingly.

Reply:
Could do. I have been trying to accomplish a couple of things here:
- reduce what subclasses need to know/implement as much as possible
- ensure rules like semantic order and paint order are followed

For the first, we can definitely put more responsibility on the subclasses to do more here, for the second, I am wary of allowing folks to deviate from rules like that.